sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/V1Beta3.java

```diff
@@ -694,141 +801,206 @@ public void populateDisplayData(DisplayData.Builder builder) {
-       * backing off between retries fails.
-       */
-      private void flushBatch() throws DatastoreException, IOException, InterruptedException {
-        LOG.debug("Writing batch of {} entities", entities.size());
-        Sleeper sleeper = Sleeper.DEFAULT;
-        BackOff backoff = new AttemptBoundedExponentialBackOff(MAX_RETRIES, INITIAL_BACKOFF_MILLIS);
-
-        while (true) {
-          // Batch upsert entities.
-          try {
-            CommitRequest.Builder commitRequest = CommitRequest.newBuilder();
-            for (Entity entity: entities) {
-              commitRequest.addMutations(makeUpsert(entity));
-            }
-            commitRequest.setMode(CommitRequest.Mode.NON_TRANSACTIONAL);
-            datastore.commit(commitRequest.build());
-            // Break if the commit threw no exception.
-            break;
-          } catch (DatastoreException exception) {
-            // Only log the code and message for potentially-transient errors. The entire exception
-            // will be propagated upon the last retry.
-            LOG.error("Error writing to the Datastore ({}): {}", exception.getCode(),
-                exception.getMessage());
-            if (!BackOffUtils.next(sleeper, backoff)) {
-              LOG.error("Aborting after {} retries.", MAX_RETRIES);
-              throw exception;
-            }
-          }
-        }
-        LOG.debug("Successfully wrote {} entities", entities.size());
-        entities.clear();
+  /**
+   * Returns true if a Datastore key is complete. A key is complete if its last element
+   * has either an id or a name.
+   */
+  static boolean isValidKey(Key key) {
+    List<PathElement> elementList = key.getPathList();
+    if (elementList.isEmpty()) {
+      return false;
+    }
+    PathElement lastElement = elementList.get(elementList.size() - 1);
+    return (lastElement.getId() != 0 || !lastElement.getName().isEmpty());
+  }
+
+  /**
+   * A function that constructs an upsert {@link Mutation} from an {@link Entity}.
+   */
+  @VisibleForTesting
+  static class UpsertFn extends SimpleFunction<Entity, Mutation> {
+    @Override
+    public Mutation apply(Entity entity) {
+      // Verify that the entity to write has a complete key.
+      if (!isValidKey(entity.getKey())) {
+        throw new IllegalArgumentException(
+            String.format("Entities to be written to the Datastore must have complete keys:\n%s",
+                entity));
       }
+      return makeUpsert(entity).build();
+    }
 
-      @Override
-      public void populateDisplayData(Builder builder) {
-        super.populateDisplayData(builder);
-        builder
-            .addIfNotNull(DisplayData.item("projectId", projectId)
-                .withLabel("Output Project"));
+    @Override
+    public void populateDisplayData(Builder builder) {
+      builder.add(DisplayData.item("upsertFn", this.getClass().getName())
+          .withLabel("Constructs upsert Mutation from Entity"));
```

Comment:
A couple things:
 a) Use `getClass()` rather than `getClass().getName()`; we keep track of some types for display data.
 b) The label on this is wordy relative to other labels in the SDK. Can we shorten it, perhaps "Create Upsert Function" ?